src/core/chart-registry.ts

```diff
@@ -39,53 +40,32 @@ class ChartRegistry {
      * @returns {Boolean}
      */
     public has (chart: BaseMixin): boolean {
-        for (const e in this._chartMap) {
-            if ((this._chartMap)[e].indexOf(chart) >= 0) {
-                return true;
+        for (const chartGroupName in this._chartMap) {
+            if (this._chartMap.hasOwnProperty(chartGroupName)) {
+                if (this._chartMap[chartGroupName].has(chart)) {
+                    return true;
+                }
             }
         }
         return false;
     }
-
-    /**
-     * Add given chart instance to the given group, creating the group if necessary.
-     * If no group is provided, the default group `constants.DEFAULT_CHART_GROUP` will be used.
-     * @param {Object} chart dc.js chart instance
-     * @param {String} [group] Group name
-     * @return {undefined}
-     */
-    public register (chart: BaseMixin, group: string): void {
-        const _chartMap = this._chartMap;
-        group = this._initializeChartGroup(group);
-        _chartMap[group].push(chart);
-    }
-
-    /**
-     * Remove given chart instance from the given group, creating the group if necessary.
-     * If no group is provided, the default group `constants.DEFAULT_CHART_GROUP` will be used.
-     * @param {Object} chart dc.js chart instance
-     * @param {String} [group] Group name
-     * @return {undefined}
-     */
-    public deregister (chart: BaseMixin, group: string): void {
-        group = this._initializeChartGroup(group);
-        for (let i = 0; i < (this._chartMap)[group].length; i++) {
-            if ((this._chartMap)[group][i].anchorName() === chart.anchorName()) {
-                (this._chartMap)[group].splice(i, 1);
-                break;
-            }
-        }
-    }
-
     /**
      * Clear given group if one is provided, otherwise clears all groups.
      * @param {String} group Group name
      * @return {undefined}
      */
-    public clear (group: string): void {
+    public clear (group?: string): void {
         if (group) {
-            delete (this._chartMap)[group];
+            if (this._chartMap[group]) {
+                this._chartMap[group].clear();
+                delete (this._chartMap)[group];
+            }
         } else {
+            for (const chartGroupName in this._chartMap) {
+                if (this._chartMap.hasOwnProperty(chartGroupName)) {
+                    this._chartMap[chartGroupName].clear();
+                }
+            }
```

Comment:
We've had occasional complaints about memory leaks when deregistering charts; this careful approach may help. 

On the other hand, this PR introduces a cycle, because the chart will hold a reference to the chart group. I guess that's why you are being careful!

Might be worth testing. Not sure if it could be automated, but we could at least have a page in the examples that creates and removes a lot of charts in a loop, and use dev tools to confirm it's not leaking.

Reply:
Created a task https://github.com/dc-js/dc.js/issues/1732